sympy/core/basic.py

```diff
@@ -1228,27 +1228,84 @@ def has(self, *patterns):
         False
 
         """
-        return any(self._has(pattern) for pattern in patterns)
+        from .traversal import iterargs
+        return self._has(iterargs, *patterns)
 
-    def _has(self, pattern):
+    @cacheit
+    def has_free(self, *patterns):
+        """return True if self has object(s) ``x`` as a free expression
+        else False.
+
+        Examples
+        ========
+
+        >>> from sympy import Integral, Function
+        >>> from sympy.abc import x, y
+        >>> f = Function('f')
+        >>> g = Function('g')
+        >>> expr = Integral(f(x), (f(x), 1, g(y)))
+        >>> expr.free_symbols
+        {y}
+        >>> expr.has_free(g(y))
+        True
+        >>> expr.has_free(*(x, f(x)))
+        False
+
+        This works for subexpressions and types, too:
+
+        >>> expr.has_free(g)
+        True
+        >>> (x + y + 1).has_free(y + 1)
+        True
+
+        """
+        from .traversal import iterfreeargs
+        return self._has(iterfreeargs, *patterns)
+
+    def _has(self, iterargs, *patterns):
+        from sympy.utilities.iterables import sift
+        from sympy.core.expr import Expr
+        from sympy.core.function import UndefinedFunction
+        from sympy.core.basic import BasicMeta
```

Comment:
Why do these things need to be special cased?

It should be possible to have a clean design that does not need special cases. It shouldn't be necessary to distinguish Expr or UndefinedFunction, This operates at the Basic level and so there are only three cases to consider:

- Explicit Basic instances
- Basic instances with Wild
- Basic subclasses

Reply:
Expressions are node arguments, classes/UndefinedFunction (except in the case of a Tuple holding UndefinedFunctions) are not: traversing args will expose Expr; traversing args and emitting `arg.func` will expose the classes. Expr and UndefinedFunction can be matched with equality whereas others might be subclasses of the classes being sought.

Symbols (not special cased) are located at the tips of branches; no comparison need be done until you get to the no-arg terminus of a branch. Doesn't it make sense to look where you are likely to find something? e.g. there is no point in looking at `arg.func` if you are looking for Expr and vice versa.

Reply:
An UndefinedFunction is a subclass of Basic so it can be handled just like any other subclass e.g. exp. I don't think that func should be used at all here. I showed efficient and simple ways to handle either expressions (Basic instances) or classes (Basic subclasses) here: https://github.com/sympy/sympy/issues/20809#issuecomment-830783820
The only other thing needed is the fallback slow path for Wild pattern matches.

Reply:
The `_has_matcher` does not have to do with Wild patterns -- it is not `match` based; it is separating and expr into commutative and noncommutative parts. That is why I try to avoid that be separating out expressions that *won't* need the matcher. 

I am happy with the performance relative to master.
```
# PR

>>> from sympy.abc import x, y, z, t
>>> expr = 1+cos(t)
>>> for i in range(10):
...     expr = 1 + expr**2 + expr
...
>>> x10 = symbols('x:10')
>>> x100 = symbols('x:100')
>>> t1=time();ok=expr.has(x);time()-t1
0.009001016616821289
>>> t1=time();ok=expr.has(t);time()-t1
0.0
>>> t1=time();ok=expr.has(cos);time()-t1
0.0
>>> t1=time();ok=expr.has(sin);time()-t1
0.017002883911132812
>>> t1=time();ok=expr.has(*x10);time()-t1
0.009001016616821289
>>> t1=time();ok=expr.has(*x100);time()-t1
0.011001110076904297


#master

>>> t1=time();ok=expr.has(x);time()-t1
0.07800793647766113
>>> t1=time();ok=expr.has(t);time()-t1
0.0
>>> t1=time();ok=expr.has(cos);time()-t1
0.0
>>> t1=time();ok=expr.has(sin);time()-t1
0.031003236770629883
>>> t1=time();ok=expr.has(*x10);time()-t1
0.7810781002044678
>>> t1=time();ok=expr.has(*x100);time()-t1
7.705770492553711
```
If you prefer the `has_expression` and `has_type` interface, that has the advantage of making the user decide which path they want and then the same sifting is not needed. But when you are dealing with BasicMeta, the user might not know whether to look for the BasicMeta like `Function('f')` as a literal `Tuple(f, g)` or as an instance `f(x) + g(y)`.

At this point in this effort, it seems to me that the use of `has` for most of the cases where `free_symbols` is used is the better thing to do. There are over 700 uses of `free_symbols` (outside of test assertions) and many of them could probably be recast using `has_free`. Reviewing them would be a good "sprint" project.

I would prefer to commit this as is (per sympy.this). I've battled with the tests to keep benchmarks from increasing. I am happy to yield to any other implementation that improves upon this while doing the same.

Reply:
I think I got confused by the use of the word match in has. I see that it doesn't do Wild patterns.

Let me have a go at writing an alternative version of `has` so we can compare them.

Reply:
> I would prefer to commit this as is (per sympy.this "commit now", principle).

To be clear that principle holds no weight and it is in fact an extremely bad idea to have a principle like that in a large software project that is used by many people downstream. We absolutely should think carefully about things before making changes because we always need to be careful about backwards compatibility. There are two sides to backwards compatibility:
1. Trying to avoid making breaking changes now
2. Trying not to create situations where breaking changes would be needed in future

Almost all discussions about "backwards compatibility" in SymPy focus on the first of these rules but the second is actually far more important. When introducing a new public method on Basic or when changing an existing method we need to be very careful to try and ensure that we will never want to change the exact semantics of that method in future. Following this principle all methods should be conservative and the possible set of valid inputs should be clearly defined.

The implementation of `has/has_free` currently in this PR is sufficiently complex (the cryptic variable names don't help) that I am not even sure what its exact semantics are. Below I show a simpler version but that fails some tests: that shows to me that the version in the PR is semantically different so that it is clear to me that it has behaviour that I would not expect even if it isn't clear exactly how they differ.

I would prefer something simpler and clearer like this:
```python
    def _has(self, iterargs, patterns):
        patterns_instance = set()
        patterns_class = []
        patterns_matcher = []
        for p in patterns:
            if isinstance(p, type):
                assert issubclass(p, Basic)
                patterns_class.append(p)
            else:
                p = _sympify(p)
                get_matcher = getattr(p, '_has_matcher', None)
                if get_matcher is not None:
                    patterns_matcher.append(get_matcher())
                else:
                    patterns_instance.add(p)

        if patterns_instance:
            if any(arg in patterns_instance for arg in iterargs(self)):
                return True

        if patterns_class:
            types = tuple(patterns_class)
            types_found = set(map(type, iterargs(self)))
            if any(issubclass(t, types) for t in types_found):
                return True

        if patterns_matcher:
            # This is the slow path (to be avoided where possible):
            for arg in iterargs(self):
                for matcher in patterns_matcher:
                    if matcher(arg):
                        return True

        return False

# Also delete the Basic _has_matcher method
```
I can see that this version fails a couple of tests. This test should be changed:
https://github.com/sympy/sympy/blob/ed0bfd1e47345c8bd45ef542eef18f6a21c55685/sympy/core/tests/test_expr.py#L1129-L1131
Since `f` and `g` are not `Basic` instances `Tuple(f, g)` should really raise an error. In any case it should not be necessary for `has/has_free` to support broken Basic instances like this.

The other test that fails is this one:
https://github.com/sympy/sympy/blob/ed0bfd1e47345c8bd45ef542eef18f6a21c55685/sympy/core/tests/test_power.py#L257-L258
The problem is something to do with `exp(x)` as a function vs `E**x` as a power. That possibly needs some special case handling via `_has_matcher` in `Pow` and/or `exp` but ideally we should avoid `_has_matcher` where possible as it takes us back down the slow path. Probably there is a bug somewhere that causes us to have a `Pow` when we should really have `exp` and that bug should be fixed rather than using `_has_matcher`.

Reply:
> Tuple(f, g) should really raise an error

I don't see a good reason to disallow a user to use a SymPy container to hold anything defined by SymPy.

Please have another look at the code. The approach is `has` or `has_free` selects the appropriate traversal iterator and calls `_has` which for Expr will isolate easy expressions to deal with before passing off to the Basic version which does two things: checks for BasicMeta and then uses a matcher which defaults to literal equality testing or (in the case of `operations._has_matcher` uses subset testing and is commutative/noncommutative aware in doing so.

Reply:
> We absolutely should think carefully about things before making changes

I made the suggestion because I thought the publicly exposed addition, `has_free`, is needed and it's signature matches that of `has` -- everything else is a detail that we are free to work out. `iterargs` and `iterfreeargs` are light versions of `preorder_traversal` and are straightforward.

Reply:
> ideally we should avoid _has_matcher where possible

That's why I sift out the args that don't need a matcher. I don't see a similar guard in the suggestion that you made. And I think that `pattern_instance` will be empty since `Basic._has_matcher` is defined and defaults to an equality tests.

Reply:
> problem is something to do with exp(x) as a function vs E**x as a power

```python
>>> from sympy.core.traversal import iterargs
>>> for i in iterargs(Pow(E, x, evaluate=0)):
...  print(i, isinstance(i, exp))
...
E**x True
E False
x False
>>> isinstance(exp(x), exp)
True
```

Your suggestion doesn't identify this because the type of `E**x` is `Pow`, not `exp`. I'm not sure where the magic methods are located that allow `isinstance` to say that the `Pow` with base `E` is to be treated like `exp`.

Reply:
> magic methods
It's defined in `__instancecheck`:
```python
class ExpMeta(FunctionClass):
    def __instancecheck__(cls, instance):
        if exp in instance.__class__.__mro__:
            return True
        return isinstance(instance, Pow) and instance.base is S.Exp1
```